### PR TITLE
fix: Encode fileServerFolder to fix #25839

### LINKED ISSUE #25839: [ERR_INVALID_CHAR]: Invalid character in header content ["x-cypress-file-path"]
### Current behavior

https://github.com/cypress-io/cypress/issues/25273 related to this closed issue and I can confirm this is happening and it's pretty annyoing. 

Error Code pops 
```
TypeError [ERR_INVALID_CHAR] [ERR_INVALID_CHAR]: Invalid character in header content ["x-cypress-file-path"]
    at ServerResponse.setHeader (_http_outgoing.js:561:3)
    at onRequest (C:\Users\Nutcracker52\AppData\Local\Cypress\Cache\8.3.1\Cypress\resources\app\packages\server\lib\file_server.js:37:7)
    at Server.<anonymous> (C:\Users\Nutcracker52\AppData\Local\Cypress\Cache\8.3.1\Cypress\resources\app\packages\server\lib\file_server.js:57:16)
    at Server.emit (events.js:315:20)
    at parserOnIncoming (_http_server.js:874:12)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:126:17)
```
  test I am trying to run 
```js
  it('validation email sent and user can confirm', () => {
    cy.getEmail(Cypress.env('emailAddress'), 'Welcome to ABC', true).then(email => {
      expect(email.subject).equal('Welcome to ABC')
      const validateUrl = getEmailLink(email.html, `${Cypress.env('appUrl')}/account/verify?code`, '"')
      cy.visit(validateUrl)
      cy.contains('you are now ready to login!', { timeout: 30000 })
    })
  })
```
  I am reading emails from an API and check that email has been received and then email contains a link and click that link
  

### Desired behavior

Cypress shouldn't throw any errors while test is running. This completely crashes Cypress and I have to run tests again. 

To avoid the error I had to add cy.visit command as first line to the code. which is not necessary. 

### Test code to reproduce

please use the given code above. 
get email is a function that we designed to get emails from an api. 
I cannot share it. 

### Cypress Version

8.3.1

### Node version

16.19.0

### Operating System

Windows 11

### Debug Logs

_No response_

### Other

_No response_

### PR DESCRIPTION
**Pull request by Zzet-Z**
- Closes https://github.com/cypress-io/cypress/issues/25839 

### Additional details
When there are invalid header characters in a test folder path (such as Chinese characters), setting the `x-cypress-file-path` header fails. This PR will encode the file path header value to solve the issue and decode it in `server-base.ts`.

### Steps to test
* Create a project that is under a folder that contains character not allowed in a header.
* Run a test that `cy.visit`s a file served from the local file server
* Verify the visit succeeds 

### How has the user experience changed?
The `Invalid character in header content ["x- cypress-file-path"]` error is no longer thrown.

### PR Tasks
- [x] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by jennifer-shehane — DISMISSED** [review]
Thanks for the contribution! Could you add a test around this behavior? We'll need a test in order to merge the fix.

**Comment by Zzet-Z** [comment]
@jennifer-shehane 
i'd be happy to do that，and I added a new unit test case, but since this issue will only occur in Ubuntu, I just can write a passing case. Is this okay?

**Comment by Zzet-Z** [comment]
@jennifer-shehane Thanks for your review！
Sorry，I hit the Unassign button by mistake😖.
This is my first time making an open source contribution. I have checked many related documents on GitHub, but I am still a little confused about the current process. Is there anything else that I need to deal with now? Looking forward to your help.

File: packages/server/test/unit/config_spec.js [hunk 1]
```diff
@@ -368,6 +368,12 @@ describe('lib/config', () => {
             return this.expectValidationFails('the value was: true')
           })
         })
+
+        it('passes if a string contain invalid character but encode it', function () {
```
Comment (jennifer-shehane) [inline packages/server/test/unit/config_spec.js]:
```suggestion
        it('passes if a string contains an invalid character but encodes it', function () {
```

**Comment by Zzet-Z** [comment]
@jennifer-shehane I'm a bit confused right now, sorry. Is there anything else need me to do,  or just wait for you to verify and initiate the merge?

**Comment by jennifer-shehane** [comment]
@Zzet-Z Just waiting til next sprint to see if I can assign some more eyes to review this.

**Comment by Zzet-Z** [comment]
> @Zzet-Z Just waiting til next sprint to see if I can assign some more eyes to review this.

OK thx！

**Review by jennifer-shehane — DISMISSED** [review]
@Zzet-Z There appear to be some failing tests with this change: https://app.circleci.com/pipelines/github/cypress-io/cypress/61086/workflows/6745104d-7a56-4c94-a750-d7c5a415b06e/jobs/2539954

**Comment by Zzet-Z** [comment]
> @Zzet-Z There appear to be some failing tests with this change: https://app.circleci.com/pipelines/github/cypress-io/cypress/61086/workflows/6745104d-7a56-4c94-a750-d7c5a415b06e/jobs/2539954

@jennifer-shehane Oh, that's a problem, and I found some other problems and dealt with them. The local operation passed these tests. Could you please help me run the integration test again if it is convenient.

**Comment by jennifer-shehane** [comment]
@Zzet-Z Running them now

**Comment by Zzet-Z** [comment]
@jennifer-shehane Oops,I neglected a test case, now it is updated, please help run it again when you are free.By the way, there are still a small part of failures that seem not to be caused by my changes, could you please help me check whether these need my attention, thx!

File: cli/CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ _Released 5/21/2024 (PENDING)_
 **Bugfixes:**
 
 - Fixed an issue where Cypress was unable to search in the Specs list for files or folders containing numbers. Fixes [#29034](https://github.com/cypress-io/cypress/issues/29034).
+- Fixed an issue parsing the `fileServerFolder` path when there are special characters. Fixes [#25839](https://github.com/cypress-io/cypress/issues/25839).
```
Comment (mschile) [inline cli/CHANGELOG.md]:
```suggestion
- Fixed an issue setting the `x-cypress-file-path` header when there are invalid header characters in the file path. Fixes [#25839](https://github.com/cypress-io/cypress/issues/25839).
```

**Review by mschile — APPROVED** [review]

**Review by ryanthemanuel — APPROVED** [review]

**Pull request merged**